src/particles.zig

```diff
@@ -0,0 +1,408 @@
+		textureIDs.deinit();
+		textures.deinit();
+		emissionTextures.deinit();
+		arenaForWorld.deinit();
+		textureArray.deinit();
+		emissionTextureArray.deinit();
+		particleTypeHashmap.deinit();
+		ParticleSystem.deinit();
+	}
+
+	pub fn register(_: []const u8, id: []const u8, zon: ZonElement) void {
+		std.log.debug("Registered particle: {s}", .{id});
+		var particleType: ParticleType = undefined;
+
+		particleType.animationFrames = zon.get(u32, "animationFrames", 1);
+
+		const textureId = zon.get([]const u8, "texture", "cubyz:spark");
+		var splitter = std.mem.splitScalar(u8, textureId, ':');
+		const mod = splitter.first();
+		const _id = splitter.rest();
+
+		// this is so confusing i just hardcoded that thing
+		const path = std.fmt.allocPrint(main.stackAllocator.allocator, "assets/{s}/particles/textures/{s}.png", .{mod, _id}) catch unreachable;
+		defer main.stackAllocator.free(path);
+		textureIDs.append(arenaForWorld.allocator().dupe(u8, path));
+		readTextureData(path, &particleType);
+
+		particleTypeHashmap.put(id, @intCast(types.items.len)) catch unreachable;
+		types.append(particleType);
+	}
+
+	fn readTextureData(_path: []const u8, typ: *ParticleType) void {
+		const animationFrames = typ.animationFrames;
+		const path = _path[0 .. _path.len - ".png".len];
+
+		const base = readTextureFile(path, ".png", Image.defaultImage);
+		const emission = readTextureFile(path, "_emission.png", Image.emptyImage);
+		typ.startFrame = @intCast(textures.items.len);
+		typ.size = @as(f32, @floatFromInt(base.width))/16;
+		for(0..animationFrames) |i| {
+			textures.append(extractAnimationSlice(base, i, animationFrames));
+			emissionTextures.append(extractAnimationSlice(emission, i, animationFrames));
+		}
+	}
+
+	fn extendedPath(_allocator: main.heap.NeverFailingAllocator, path: []const u8, ending: []const u8) []const u8 {
+		return std.fmt.allocPrint(_allocator.allocator, "{s}{s}", .{path, ending}) catch unreachable;
+	}
+
+	fn readTextureFile(_path: []const u8, ending: []const u8, default: Image) Image {
+		const path = extendedPath(main.stackAllocator, _path, ending);
+		defer main.stackAllocator.free(path);
+		return Image.readFromFile(arenaForWorld.allocator(), path) catch default;
+	}
+
+	fn extractAnimationSlice(image: Image, frame: usize, frames: usize) Image {
+		if(image.height < frames) return image;
+		var startHeight = image.height/frames*frame;
+		if(image.height%frames > frame) startHeight += frame else startHeight += image.height%frames;
+		var endHeight = image.height/frames*(frame + 1);
+		if(image.height%frames > frame + 1) endHeight += frame + 1 else endHeight += image.height%frames;
+		var result = image;
+		result.height = @intCast(endHeight - startHeight);
+		result.imageData = result.imageData[startHeight*image.width .. endHeight*image.width];
+		return result;
+	}
+
+	pub fn generateTextureArray() void {
+		std.log.debug("Particle texture sizes: {d} {d}", .{textures.items.len, emissionTextures.items.len});
```

Comment:
No need to print two copies of the same number.
If you want to check if they are the same, use an assertion.